class-memaccess warning when compiling with GCC 8 or later

Building Sparsehash trunk or 2.0.3 from source with GCC 8.2 gives various `class-memaccess` warnings like these:

```
./src/sparsehash/internal/libc_allocator_with_realloc.h:68:40: warning:
‘void* realloc(void*, size_t)’ moving an object of non-trivially copyable type
‘class std::__cxx11::basic_string<char>’; use ‘new’ and ‘delete’ instead [-Wclass-memaccess]
     return static_cast<pointer>(realloc(p, n * sizeof(value_type)));
                                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
./src/sparsehash/sparsetable:1157:13: warning:
‘void* memcpy(void*, const void*, size_t)’ writing to an object of type
‘struct std::pair<const int, {anonymous}::Memmove>’ with no trivial copy-assignment;
use copy-initialization instead [-Wclass-memaccess]
       memcpy(group + i, group + i+1, sizeof(*group));  // hopefully inlined!
       ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

A regular allocator (e.g. `std::allocator`) must be used to avoid undefined behavior unless `is_trivially_copyable<value_type>` is true.  As Sparsehash is written today, the default is a custom allocator using `realloc()` which invokes undefined behavior for many common types, including `std::string` as either key or value type.

Discussion: https://stackoverflow.com/questions/52436264/google-sparsehash-uses-realloc-on-type-which-is-not-trivially-copyable

--- a/src/sparsehash/internal/libc_allocator_with_realloc.h
+++ b/src/sparsehash/internal/libc_allocator_with_realloc.h
@@ -65,7 +65,10 @@ class libc_allocator_with_realloc {
     free(p);
   }
   pointer reallocate(pointer p, size_type n) {
-    return static_cast<pointer>(realloc(p, n * sizeof(value_type)));
+    // p points to a storage array whose objects have already been destroyed
+    // cast to void* to prevent compiler warnings about calling realloc() on
+    // an object which cannot be relocated in memory
+    return static_cast<pointer>(realloc(static_cast<void*>(p), n * sizeof(value_type)));
   }
 
   size_type max_size() const  {

